File: backend/routers/contracts.py
from fastapi import APIRouter, UploadFile, File, HTTPException, Depends
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session
from typing import List, Dict, Any
import io
import pandas as pd

from backend.database import get_db
from backend.schemas.contract import ContractResponse, ContractCreate
from backend.services.file_service import extract_text_from_file
from backend.services.ai_service import extract_contract_details
from backend.services.contract_service import (
    create_contract,
    get_user_contracts,
    get_contract_by_id,
    delete_contract_by_id
)
from backend.utils.auth import get_current_user
from backend.models.user import User

router = APIRouter(prefix="/api/contracts", tags=["contracts"])

@router.options("/extract")
async def extract_options():
    """Handle CORS preflight for extract endpoint."""
    return {}

@router.post("/extract")
async def extract_contract(file: UploadFile = File(...)):
    """Extract contract details from file without saving to database."""
    try:
        # Validate file type
        allowed_extensions = ['pdf', 'docx', 'doc']
        file_extension = file.filename.lower().split('.')[-1]

        if file_extension not in allowed_extensions:
            raise HTTPException(
                status_code=400,
                detail=f"Invalid file type. Allowed types: {', '.join(allowed_extensions)}"
            )

        # Read file content
        content = await file.read()

        # Extract text from file
        text = extract_text_from_file(content, file.filename)

        # Extract contract details using AI
        details = await extract_contract_details(text)

        # Return extracted details with file name
        return {
            "file_name": file.filename,
            **details
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/upload", response_model=ContractResponse)
async def upload_contract(
    contract_data: ContractCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Save contract with provided data to database."""
    try:
        # Create contract in database with provided data and current user
        contract = create_contract(
            db,
            contract_data.dict(exclude={'file_name'}),
            contract_data.file_name,
            current_user.id
        )

        return contract
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail=str(e))

@router.get("", response_model=List[ContractResponse])
def get_contracts(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get all contracts for the current user."""
    return get_user_contracts(db, current_user.id)

@router.get("/{contract_id}", response_model=ContractResponse)
def get_contract(
    contract_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get a specific contract by ID for the current user."""
    contract = get_contract_by_id(db, contract_id, current_user.id)
    if not contract:
        raise HTTPException(status_code=404, detail="Contract not found")
    return contract

@router.delete("/{contract_id}")
def delete_contract(
    contract_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Delete a contract by ID for the current user."""
    success = delete_contract_by_id(db, contract_id, current_user.id)
    if not success:
        raise HTTPException(status_code=404, detail="Contract not found")
    return {"message": "Contract deleted successfully"}

@router.get("/export/csv")
def export_contracts_csv(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Export all contracts for the current user to CSV."""
    contracts = get_user_contracts(db, current_user.id)
    
    # Convert to DataFrame
    data = [{
        "ID": c.id,
        "File Name": c.file_name,
        "Contact Name": c.contact_name,
        "Contact Email": c.contact_email,
        "Contact Phone": c.contact_phone,
        "Start Date": c.start_date,
        "End Date": c.end_date,
        "Contract Value": c.contract_value,
        "Payment Terms": c.payment_terms,
        "Termination Terms": c.termination_terms,
        "Summary": c.summary,
        "Created At": c.created_at
    } for c in contracts]
    
    df = pd.DataFrame(data)
    
    # Create CSV in memory
    stream = io.StringIO()
    df.to_csv(stream, index=False)
    stream.seek(0)
    
    return StreamingResponse(
        iter([stream.getvalue()]),
        media_type="text/csv",
        headers={"Content-Disposition": "attachment; filename=contracts.csv"}
    )

File: src/lib/contracts.ts
import { supabase } from './supabase';

export interface Contract {
  id: number;
  user_id: string;
  file_name: string;
  contact_name: string | null;
  contact_email: string | null;
  contact_phone: string | null;
  start_date: string | null;
  end_date: string | null;
  contract_value: number | null;
  payment_terms: string | null;
  termination_terms: string | null;
  summary: string | null;
  created_at: string;
}

export interface ExtractedContract {
  file_name: string;
  contact_name: string | null;
  contact_email: string | null;
  contact_phone: string | null;
  start_date: string | null;
  end_date: string | null;
  contract_value: number | null;
  payment_terms: string | null;
  termination_terms: string | null;
  summary: string | null;
}

const API_BASE_URL = 'http://localhost:8000';

export const contractsService = {
  // Extract contract using backend AI service
  async extractContract(file: File): Promise<ExtractedContract> {
    const formData = new FormData();
    formData.append('file', file);

    const response = await fetch(`${API_BASE_URL}/api/contracts/extract`, {
      method: 'POST',
      body: formData,
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to extract contract details');
    }

    return response.json();
  },

  // Save contract to Supabase
  async saveContract(contractData: ExtractedContract): Promise<Contract> {
    const { data: { session } } = await supabase.auth.getSession();

    if (!session) {
      throw new Error('Not authenticated');
    }

    const { data, error } = await supabase
      .from('contracts')
      .insert([
        {
          user_id: session.user.id,
          ...contractData,
        },
      ])
      .select()
      .single();

    if (error) throw error;
    return data;
  },

  // Get all contracts for current user
  async getContracts(): Promise<Contract[]> {
    const { data: { session } } = await supabase.auth.getSession();

    if (!session) {
      throw new Error('Not authenticated');
    }

    const { data, error } = await supabase
      .from('contracts')
      .select('*')
      .eq('user_id', session.user.id)
      .order('created_at', { ascending: false });

    if (error) throw error;
    return data || [];
  },

  // Get a single contract
  async getContract(id: number): Promise<Contract> {
    const { data: { session } } = await supabase.auth.getSession();

    if (!session) {
      throw new Error('Not authenticated');
    }

    const { data, error } = await supabase
      .from('contracts')
      .select('*')
      .eq('id', id)
      .eq('user_id', session.user.id)
      .single();

    if (error) throw error;
    return data;
  },

  // Delete a contract
  async deleteContract(id: number): Promise<void> {
    const { data: { session } } = await supabase.auth.getSession();

    if (!session) {
      throw new Error('Not authenticated');
    }

    const { error } = await supabase
      .from('contracts')
      .delete()
      .eq('id', id)
      .eq('user_id', session.user.id);

    if (error) throw error;
  },

  // Export contracts to CSV
  async exportToCSV(): Promise<Blob> {
    const contracts = await this.getContracts();

    // Convert contracts to CSV
    const headers = [
      'ID',
      'File Name',
      'Contact Name',
      'Contact Email',
      'Contact Phone',
      'Start Date',
      'End Date',
      'Contract Value',
      'Payment Terms',
      'Termination Terms',
      'Summary',
      'Created At',
    ];

    const csvRows = [
      headers.join(','),
      ...contracts.map(contract =>
        [
          contract.id,
          `"${contract.file_name}"`,
          `"${contract.contact_name || ''}"`,
          `"${contract.contact_email || ''}"`,
          `"${contract.contact_phone || ''}"`,
          `"${contract.start_date || ''}"`,
          `"${contract.end_date || ''}"`,
          contract.contract_value || '',
          `"${contract.payment_terms || ''}"`,
          `"${contract.termination_terms || ''}"`,
          `"${contract.summary || ''}"`,
          `"${contract.created_at}"`,
        ].join(',')
      ),
    ];

    const csvContent = csvRows.join('\n');
    return new Blob([csvContent], { type: 'text/csv' });
  },

  // Chat analytics (still uses backend for AI processing)
  async chatAnalytics(message: string): Promise<string> {
    const { data: { session } } = await supabase.auth.getSession();

    if (!session) {
      throw new Error('Not authenticated');
    }

    const response = await fetch(`${API_BASE_URL}/api/analytics/chat`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${session.access_token}`,
      },
      body: JSON.stringify({ message }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to process chat');
    }

    const data = await response.json();
    return data.response;
  },
};
